File: ChatBotFAQ/app/services/llm.py
"""Build the grounded prompt and stream the model's answer (Gemini or Gemma,
both served through the same google-genai client).

The retrieved passages are passed as a numbered list so the model can cite by
number, and the answering rules go in the system instruction rather than being
repeated per turn — that keeps the per-question part of the request small.

`raw` is what reaches the model, not `content`: the breadcrumb prefix on
`content` exists to steer the *embedding*, and feeding it to the model would
just be noise it might quote back.
"""

import os
import sys
import time
from pathlib import Path

from google import genai
from google.genai import types

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

MODEL = os.environ.get("CHATBOT_MODEL", "gemini-3.1-flash-lite")
MAX_OUTPUT_TOKENS = 2048

# Thinking budgets are a Gemini-only feature (Gemma has no "thinking" mode), so
# this is only read/applied when MODEL is a Gemini model — see build_config().
# Kept for whoever switches back: on gemini-3.5-flash-lite the same question
# took 2.0s at MINIMAL and 148.8s at LOW, for answers that differed only in
# word choice, so MINIMAL was the right default there too.
THINKING_LEVEL = os.environ.get("CHATBOT_THINKING", "MINIMAL").upper()

# Give up rather than leave the browser watching an open stream that will never
# produce a token. Free-tier latency is wildly uneven — the same question was
# measured at 2.0s and at 148.8s — and from the user's side a slow call and a
# dead one look the same, so the only difference that matters is whether they
# are told.
TIMEOUT_MS = int(os.environ.get("CHATBOT_TIMEOUT_MS", "10000"))

# Free-tier requests per minute is per project *per model* and varies by model
# — check the current quota for MODEL on the AI Studio/API quota page and set
# CHATBOT_RPM to match. Nothing enforces it here — it is quoted so the 429
# message tells the user a number that matches their plan, and so the
# evaluation script has one place to read it from.
FREE_TIER_RPM = int(os.environ.get("CHATBOT_RPM", "15"))

# How many bullets the answer may use. Kept small on purpose: the failure mode
# of a grounded assistant is padding a thin retrieval into a full-looking answer.
MAX_POINTS = int(os.environ.get("CHATBOT_MAX_POINTS", "5"))

# The exact string the model must return when the passages do not answer the
# question. Fixed and short so it can be detected downstream, and so there is no
# room to soften a refusal into a guess.
NO_ANSWER = (
    "Tôi chưa rõ câu hỏi của bạn, bạn có thể đặt ra câu hỏi chi tiết hơn được không ạ?"
)
NOT_UPDATED = "Xin lỗi, tôi chưa được cập nhật thông tin mới nhất. Bạn có thể tham khảo các nguồn chính thức hoặc liên hệ trực tiếp với BKFintech để biết thông tin chi tiết."

SYSTEM_PROMPT = f"""\
Bạn là trợ lý BKFintech (Viện Công nghệ và Kinh tế số, ĐH Bách khoa Hà Nội). \
Chỉ dùng thông tin trong <data>. Không dùng kiến thức ngoài, không tra cứu \
ngoài, không suy luận thêm những gì văn bản không nói.
Văn phong:
- Xưng "mình", gọi người dùng là "bạn";
- Nói như trò chuyện. Mở đầu ngắn ("Dạ,") được; câu dẫn thủ tục \
("Dựa trên thông tin được cung cấp…", "Theo tài liệu…") không.
- Không chấm than, không nịnh, không xin lỗi dài dòng.
- <data> không đủ trả lời câu hỏi → trả lời đúng một câu, không gì khác: \
"{NOT_UPDATED}"
- nếu như bạn không hiểu câu hỏi của user hoặc phạm vi của câu hỏi quá rộng, hãy trả lời đúng một câu: "{NO_ANSWER}"
- <data> đúng đối tượng hỏi (khóa học, chương trình...) nhưng THIẾU chi tiết \
câu hỏi cần (chi phí, thời lượng, ngày khai giảng, năm thành lập...) → trả lời "{NOT_UPDATED}" \
kèm số nguồn xác nhận đối tượng. KHÔNG dùng câu "{NO_ANSWER}" cho trường hợp này.
- Trả lời thẳng vào vấn đề, không dẫn kiểu "Dựa vào văn bản/Theo thông tin cung cấp".
- Tối đa {MAX_POINTS} gạch đầu dòng; câu đơn giản thì 1 ý là đủ.
- Không bịa ngày tháng, số liệu, tên người, giá tiền ngoài <data>.
- Trả lời đúng ngôn ngữ câu hỏi; nguồn khác ngôn ngữ thì dịch phần cần dùng, \
không đổi ngôn ngữ trả lời, không xin lỗi vì điều đó.
Văn phong:
Xưng "mình", gọi người dùng là "bạn";
Nói như trò chuyện. Mở đầu ngắn ("Dạ,") được; câu dẫn thủ tục \
("Dựa trên thông tin được cung cấp…", "Theo tài liệu…") không.
Không chấm than, không nịnh, không xin lỗi dài dòng.
- Nguồn mâu thuẫn nhau → nêu cả hai, chỉ rõ khác biệt, không tự chọn một bên.
- Lượt hỏi-đáp trước chỉ dùng khi câu hỏi hiện tại phụ thuộc ngữ cảnh (đại từ, \
hỏi tiếp điều vừa nhắc). Chủ đề mới, độc lập → bỏ qua lượt trước.\
"""


def format_sources(chunks: list[dict]) -> str:
    blocks = []
    for i, chunk in enumerate(chunks, start=1):
        title = (chunk.get("title") or "").strip()
        published = chunk.get("published_at")
        header = f"[{i}] {title}" if title else f"[{i}]"
        if published:
            header += f" ({published})"
        blocks.append(f"{header}\n{chunk['url']}\n{chunk['raw']}")
    return "\n\n".join(blocks)


def build_user_message(question: str, chunks: list[dict]) -> str:
    """Question first, then the passages inside an explicit <data> boundary.

    The delimiter is doing real work: it marks exactly where quotable material
    starts and stops, so "only what is inside <data>" is a rule the model can
    actually check itself against, rather than a vague instruction about
    "the sources".
    """
    if not chunks:
        # Nothing retrieved is itself an empty <data> block, so the NO_ANSWER
        # rule applies on its own without a second code path here.
        return f"Câu hỏi: {question}\n\n<data>\n</data>"
    return f"Câu hỏi: {question}\n\n" "<data>\n" f"{format_sources(chunks)}\n" "</data>"


def build_contents(question: str, chunks: list[dict], history: list[dict]) -> list:
    """Prior turns plus the current question, as the multi-turn list the SDK
    expects — not one string with the history flattened into it, which would
    leave the model unable to tell where a past answer ends and the live
    question's <data> block begins.

    `history` stores roles as "user"/"assistant" (see chat_history.py); Gemini
    calls the model's own turns "model", so that rename happens only here, at
    the boundary, rather than leaking the Gemini-specific name into storage.
    """
    turns = [
        types.Content(
            role="model" if msg["role"] == "assistant" else "user",
            parts=[types.Part(text=msg["content"])],
        )
        for msg in history
    ]
    turns.append(
        types.Content(
            role="user",
            parts=[types.Part(text=build_user_message(question, chunks))],
        )
    )
    return turns


def build_config() -> types.GenerateContentConfig:
    kwargs = dict(
        system_instruction=SYSTEM_PROMPT,
        max_output_tokens=MAX_OUTPUT_TOKENS,
    )
    # Gemma rejects thinking_config outright — it's a Gemini-only field, so it
    # is only attached when MODEL is actually a Gemini model.
    if MODEL.startswith("gemini"):
        kwargs["thinking_config"] = types.ThinkingConfig(thinking_level=THINKING_LEVEL)
    return types.GenerateContentConfig(**kwargs)


def build_request(
    question: str, chunks: list[dict], history: list[dict] = None
) -> dict:
    """What gets sent to the API, as a plain dict for preview_prompt.py.

    Kept separate from `build_config()` so the prompt can be inspected without
    constructing SDK objects or holding an API key.
    """
    config = {"max_output_tokens": MAX_OUTPUT_TOKENS}
    if MODEL.startswith("gemini"):
        config["thinking_level"] = THINKING_LEVEL
    return {
        "model": MODEL,
        "system": SYSTEM_PROMPT,
        "contents": build_contents(question, chunks, history or []),
        "config": config,
    }


class MockAnswerer:
    """Stand-in for `Answerer` that never calls the API.

    Retrieval still runs for real; instead of an answer, the passages the model
    *would* have been given are printed verbatim. That makes the window the
    grounding actually reaches the model — `raw`, not `content` — readable in
    the browser. Set CHATBOT_MOCK=1 to select it.

    Matches the retriever's default top_k so the mock shows exactly what the
    model would have been given, not a sample of it.
    """

    SHOW = 7

    def stream(self, question: str, chunks: list[dict]):
        shown = chunks[: self.SHOW]
        yield (
            f"[CHẾ ĐỘ MOCK — không gọi LLM] Nội dung {len(shown)} đoạn gần "
            f"nhất với câu hỏi, trên tổng {len(chunks)} đoạn truy hồi được:\n"
        )
        for i, chunk in enumerate(shown, start=1):
            title = (chunk.get("title") or "").strip() or "(không có tiêu đề)"
            # Line by line rather than word by word: five full chunks streamed a
            # word at a time would take half a minute to finish drawing.
            for line in (
                f"\n{'─' * 60}\n"
                f"[{i}] score {chunk['score']:.4f} · {chunk.get('collection')} · {chunk['chunk_id']}\n"
                f"{title}\n{chunk['url']}\n\n{chunk['raw']}\n"
            ).splitlines(keepends=True):
                time.sleep(0.01)
                yield line


class AnswerTimeout(TimeoutError):
    pass


class Answerer:
    def __init__(self, client: genai.Client = None):
        # Resolves GEMINI_API_KEY, then GOOGLE_API_KEY, from the environment.
        # The HTTP timeout is the only one that can interrupt a call that has
        # not sent a single byte yet — a wall-clock check in the loop below
        # never runs, because the generator is still blocked inside the SDK.
        self.client = client or genai.Client(
            http_options=types.HttpOptions(timeout=TIMEOUT_MS)
        )

    def stream(self, question: str, chunks: list[dict], history: list[dict] = None):
        """Yield answer text as it is generated, or give up at the deadline.

        Streaming rather than a single response: a grounded answer over seven
        passages takes long enough that a chat window would otherwise sit blank.

        Two guards, because they catch different failures. The HTTP timeout
        handles a request that never starts; the deadline below handles one that
        starts and then trickles — measured on the free tier, the same question
        answered in 2s once and 148s another time, and the slow case looks
        identical to a hang from the browser.
        """
        deadline = time.monotonic() + TIMEOUT_MS / 1000
        for part in self.client.models.generate_content_stream(
            model=MODEL,
            contents=build_contents(question, chunks, history or []),
            config=build_config(),
        ):
            if part.text:
                yield part.text
            if time.monotonic() > deadline:
                raise AnswerTimeout(
                    f"Model không trả lời xong trong {TIMEOUT_MS / 1000:.0f} giây."
                )

File: bkft-chatbot-js/src/lib/llm.ts
/**
 * Build the grounded prompt and stream the model's answer via the Vercel AI
 * SDK (`ai` + `@ai-sdk/google`). Direct port of llm.py's prompt/streaming
 * logic; the Gemini-specific SDK call is the one substantial rewrite since
 * the Python side spoke to google-genai directly and this speaks through the
 * AI SDK's provider abstraction instead.
 *
 * `raw` is what reaches the model, not a cleaned/embedding-only field — same
 * reasoning as llm.py: any breadcrumb text added purely to steer the
 * *embedding* would just be noise for the model to quote back.
 */
import { streamText, type ModelMessage } from "ai";
import { google } from "@ai-sdk/google";
import {
  CHAT_MODEL,
  MAX_OUTPUT_TOKENS,
  MAX_POINTS,
  MOCK,
  TIMEOUT_MS,
} from "./config";
import type { RetrievalChunk } from "./retriever";
import type { ChatMessage } from "./chatHistory";

// The exact string the model must return when the passages don't answer the
// question. Fixed and short so it can be detected downstream if needed, and
// so there's no room to soften a refusal into a guess.
export const NO_ANSWER =
  "Tôi chưa rõ câu hỏi của bạn, bạn có thể đặt ra câu hỏi chi tiết hơn được không ạ?";
export const NOT_UPDATED =
  "Xin lỗi, tôi chưa được cập nhật thông tin mới nhất. Bạn có thể tham khảo các nguồn chính thức hoặc liên hệ trực tiếp với BKFintech để biết thông tin chi tiết.";

export const SYSTEM_PROMPT = `\
Bạn là trợ lý BKFintech (Viện Công nghệ và Kinh tế số, ĐH Bách khoa Hà Nội). \
Chỉ dùng thông tin trong <data>. Không dùng kiến thức ngoài, không tra cứu \
ngoài, không suy luận thêm những gì văn bản không nói.

- <data> không đủ trả lời câu hỏi → trả lời đúng một câu, không gì khác: \
"${NOT_UPDATED}"
- nếu như bạn không hiểu câu hỏi của user hoặc phạm vi của câu hỏi quá rộng, hãy trả lời đúng một câu: "${NO_ANSWER}"
- <data> đúng đối tượng hỏi (khóa học, chương trình...) nhưng THIẾU chi tiết \
câu hỏi cần (chi phí, thời lượng, ngày khai giảng, năm thành lập...) → trả lời "${NOT_UPDATED}" \
kèm số nguồn xác nhận đối tượng. KHÔNG dùng câu "${NO_ANSWER}" cho trường hợp này.
- Trả lời thẳng vào vấn đề, không dẫn kiểu "Dựa vào văn bản/Theo thông tin cung cấp".
- Tối đa ${MAX_POINTS} gạch đầu dòng, mỗi ý một dòng; câu đơn giản thì 1 ý là đủ.
- Cuối mỗi ý ghi số nguồn: [1] hoặc [2][5].
- Không bịa ngày tháng, số liệu, tên người, giá tiền ngoài <data>.
- Trả lời đúng ngôn ngữ câu hỏi; nguồn khác ngôn ngữ thì dịch phần cần dùng, \
không đổi ngôn ngữ trả lời, không xin lỗi vì điều đó.
- Nguồn mâu thuẫn nhau → nêu cả hai, chỉ rõ khác biệt, không tự chọn một bên.
- Lượt hỏi-đáp trước chỉ dùng khi câu hỏi hiện tại phụ thuộc ngữ cảnh (đại từ, \
hỏi tiếp điều vừa nhắc). Chủ đề mới, độc lập → bỏ qua lượt trước.`;

export function formatSources(chunks: RetrievalChunk[]): string {
  return chunks
    .map((chunk, i) => {
      const n = i + 1;
      const title = (chunk.title || "").trim();
      const header = title ? `[${n}] ${title}` : `[${n}]`;
      const withDate = chunk.published_at
        ? `${header} (${chunk.published_at})`
        : header;
      return `${withDate}\n${chunk.url}\n${chunk.raw}`;
    })
    .join("\n\n");
}

export function buildUserMessage(
  question: string,
  chunks: RetrievalChunk[]
): string {
  // Nothing retrieved is itself an empty <data> block, so the NOT_UPDATED /
  // NO_ANSWER rules apply on their own without a second code path — same as
  // llm.py's build_user_message.
  if (chunks.length === 0) return `Câu hỏi: ${question}\n\n<data>\n</data>`;
  return `Câu hỏi: ${question}\n\n<data>\n${formatSources(chunks)}\n</data>`;
}

/** Prior turns plus the current question, as the multi-turn message list the
 * AI SDK expects — not one flattened string, so the model can tell where a
 * past answer ends and the live question's <data> block begins. */
export function buildMessages(
  question: string,
  chunks: RetrievalChunk[],
  history: ChatMessage[]
): ModelMessage[] {
  const turns: ModelMessage[] = history.map((msg) => ({
    role: msg.role === "assistant" ? "assistant" : "user",
    content: msg.content,
  }));
  turns.push({ role: "user", content: buildUserMessage(question, chunks) });
  return turns;
}

export class AnswerTimeout extends Error {}

/** Yield answer text as it's generated, or give up at TIMEOUT_MS.
 *
 * Two guards, same as llm.py, because they catch different failures: the AI
 * SDK call itself can hang before the first byte (network/provider issue),
 * and a slow-but-live stream can trickle past the deadline — on the free
 * tier the same question was measured (Python side) at 2s once and ~150s
 * another time, and the slow case looks identical to a hang from the
 * client's side. Both are treated the same way here: stop yielding and throw.
 */
export async function* streamAnswer(
  question: string,
  chunks: RetrievalChunk[],
  history: ChatMessage[] = []
): AsyncGenerator<string> {
  const deadline = Date.now() + TIMEOUT_MS;
  const result = streamText({
    model: google(CHAT_MODEL),
    system: SYSTEM_PROMPT,
    messages: buildMessages(question, chunks, history),
    maxOutputTokens: MAX_OUTPUT_TOKENS,
  });

  for await (const delta of result.textStream) {
    yield delta;
    if (Date.now() > deadline) {
      throw new AnswerTimeout(
        `Model không trả lời xong trong ${Math.round(TIMEOUT_MS / 1000)} giây.`
      );
    }
  }
}

/** Stand-in for streamAnswer() that never calls Gemini. Retrieval still runs
 * for real; the passages the model *would* have received are echoed back
 * verbatim, line by line, so the window the grounding actually reaches the
 * model is readable in the browser. Matches llm.py's MockAnswerer, including
 * its SHOW cap of 7 to mirror the retriever's default top_k. */
const MOCK_SHOW = 7;
export async function* mockStreamAnswer(
  chunks: RetrievalChunk[]
): AsyncGenerator<string> {
  const shown = chunks.slice(0, MOCK_SHOW);
  yield (
    `[CHẾ ĐỘ MOCK — không gọi LLM] Nội dung ${shown.length} đoạn gần ` +
    `nhất với câu hỏi, trên tổng ${chunks.length} đoạn truy hồi được:\n`
  );
  for (let i = 0; i < shown.length; i++) {
    const chunk = shown[i];
    const title = (chunk.title || "").trim() || "(không có tiêu đề)";
    const block =
      `\n${"─".repeat(60)}\n` +
      `[${i + 1}] score ${chunk.score.toFixed(4)} · ${chunk.collection} · ${chunk.chunk_id}\n` +
      `${title}\n${chunk.url}\n\n${chunk.raw}\n`;
    // Line by line rather than the whole block at once, so the mock stream
    // still exercises the SSE "delta" path the same shape a real answer
    // would — same reasoning as MockAnswerer.stream()'s per-line sleep.
    for (const line of block.split(/(?<=\n)/)) {
      await new Promise((r) => setTimeout(r, 5));
      yield line;
    }
  }
}

/** Says which of the two likely failures happened, in the user's language.
 * The distinction matters: a timeout is worth retrying immediately, a quota
 * error is not — waiting is the only thing that helps. Port of llm.py's
 * friendly_error / main.py usage of it. */
export function friendlyError(err: unknown): string {
  const text = err instanceof Error ? err.message : String(err);
  const isTimeout = err instanceof AnswerTimeout || /timeout/i.test(text);
  if (isTimeout) {
    return (
      `Model không phản hồi trong ${Math.round(TIMEOUT_MS / 1000)} giây. ` +
      "Nguồn tham khảo ở trên vẫn đúng — thử hỏi lại."
    );
  }
  if (/429|RESOURCE_EXHAUSTED/.test(text)) {
    return (
      "Đã chạm giới hạn câu hỏi mỗi phút của gói miễn phí. " +
      "Đợi khoảng một phút rồi hỏi lại."
    );
  }
  const name = err instanceof Error ? err.constructor.name : "Error";
  return `${name}: ${text.slice(0, 200)}`;
}

export function answerStream(
  question: string,
  chunks: RetrievalChunk[],
  history: ChatMessage[]
): AsyncGenerator<string> {
  return MOCK ? mockStreamAnswer(chunks) : streamAnswer(question, chunks, history);
}
